[iOS] Prepare API tests for async text input enablement

whsieh opened the pull request:
#### 043e71bfb0158320e9539bfb750d7e74bd9f5d7b
<pre>
[iOS] Prepare API tests for async text input enablement
<a href="https://bugs.webkit.org/show_bug.cgi?id=266578">https://bugs.webkit.org/show_bug.cgi?id=266578</a>
<a href="https://rdar.apple.com/119816733">rdar://119816733</a>

Reviewed by Aditya Keerthi.

Prepare our API tests for async text input enablement on iOS, many of which currently call into
legacy SPI methods that crash with the feature flag enabled. To avoid these release assertions, we
refactor all tests which explicitly call into these codepaths, such that they&apos;ll invoke their
`UIAsyncTextInput` counterparts when `UIKit/async_text_input` is enabled.

See below for more details.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView moveInDirection:byGranularity:]):
(-[WKContentView extendInDirection:byGranularity:]):

Also drive-by fix an actual bug that was caught by `SelectionTests.ModifyByParagraphBoundary`.
Instead of just moving backward or forward by paragraph granularity in the `UIAsyncTextInput` method
`-extendInDirection:byGranularity:`, we need to additionally move backwards or forwards by character
before doing so in order to maintain parity with `-_moveToEndOfParagraph:withHistory:` and
`_moveToStartOfParagraph:withHistory:`.

* Source/WebKit/UIProcess/ios/WKSEDefinitions.h: Renamed from Source/WebKit/UIProcess/API/ios/WKSEDefinitions.h.

Also move `WKSEDefinitions.h` to `UIProcess/ios` instead of `UIProcess/API/ios`, since the latter
just holds iOS-specific files related to WebKit API objects (in this case, only `WKWebView`).
Putting this in `UIProcess/ios` seems more appropriate, and also allows us to include it directly in
WebKitTestRunner.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestRunnerShared/spi/UIKitSPIForTesting.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
(-[TestWKWebView synchronouslyAdjustSelectionWithDelta:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ImageAnalysisTests.mm:
(TestWebKitAPI::simulateEditContextMenuAppearance):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentViewEditingActions.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEditActions.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(TestWebKitAPI::simulateEditContextMenuAppearance):
* Tools/TestWebKitAPI/Tests/ios/AutocorrectionTestsIOS.mm:
(-[TestWKWebView autocorrectionRectsForString:]): Deleted.
(-[TestWKWebView synchronouslyRequestAutocorrectionContext]): Deleted.
* Tools/TestWebKitAPI/Tests/ios/EnterKeyHintTests.mm:
(-[TestWKWebView test:enterKeyHint:returnKeyType:]):
* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
* Tools/TestWebKitAPI/Tests/ios/SelectionModifyByParagraphBoundary.mm:
* Tools/TestWebKitAPI/Tests/ios/TextServicesTests.mm:
* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:

Add a bunch of new category helpers on `WKWebView`, which replace many of the direct calls into
`WKContentView` in our API tests when exercising various `UIWKInteractionViewProtocol` SPIs. These
helper methods will call directly into these legacy SPIs when async text input is disabled, but will
call into their new `UIAsyncTextInput` counterparts when async text input is enabled.

(-[WKWebView defineSelection]):
(-[WKWebView shareSelection]):
(-[WKWebView hasAsyncTextInput]):
(-[WKWebView selectionClipRect]):
(-[WKWebView moveSelectionToStartOfParagraph]):
(-[WKWebView extendSelectionToStartOfParagraph]):
(-[WKWebView moveSelectionToEndOfParagraph]):
(-[WKWebView extendSelectionToEndOfParagraph]):
(-[WKWebView asyncTextInput]):
(-[WKWebView extendedTextInputTraits]):
(-[WKWebView autocorrectionRectsForString:]):
(-[WKWebView autocorrectionContext]):
(-[WKWebView effectiveTextInputTraits]):
(-[WKWebView replaceText:withText:shouldUnderline:completion:]):
(wrap):
(unwrap):
(-[WKWebView synchronouslyAdjustSelectionWithDelta:]):
(-[WKWebView selectTextForContextMenuWithLocationInView:completion:]):
(-[WKWebView handleKeyEvent:completion:]):
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::asyncTextInput const):

Canonical link: <a href="https://commits.webkit.org/272233@main">https://commits.webkit.org/272233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61ae67daaba65a2649b5d1c7ddb85ca5f159cb7b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30991 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32680 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33508 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27999 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31754 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12038 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6928 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27854 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8168 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27725 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6995 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7184 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27604 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34846 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28242 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28088 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33303 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5269 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31140 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8900 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7311 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7906 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7745 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/ios/AutocorrectionTestsIOS.mm:
```diff
@@ -242,7 +213,7 @@ static void checkCGRectIsNotEmpty(CGRect rect)
     __block bool done = false;
 
-    [[webView textInputContentView] applyAutocorrection:@"tomorrow night" toString:@"tomorrownight" shouldUnderline:YES withCompletionHandler:^(UIWKAutocorrectionRects *) {
+    [webView replaceText:@"tomorrownight"  withText:@"tomorrow night" shouldUnderline:YES completion:^{
```
Extra space:
```suggestion
    [webView replaceText:@"tomorrownight" withText:@"tomorrow night" shouldUnderline:YES completion:^{
```

whsieh replied:
Good catch!

whsieh committed:
[iOS] Prepare API tests for async text input enablement
https://bugs.webkit.org/show_bug.cgi?id=266578
rdar://119816733

Reviewed by Aditya Keerthi.

Prepare our API tests for async text input enablement on iOS, many of which currently call into
legacy SPI methods that crash with the feature flag enabled. To avoid these release assertions, we
refactor all tests which explicitly call into these codepaths, such that they'll invoke their
`UIAsyncTextInput` counterparts when `UIKit/async_text_input` is enabled.

See below for more details.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView moveInDirection:byGranularity:]):
(-[WKContentView extendInDirection:byGranularity:]):

Also drive-by fix an actual bug that was caught by `SelectionTests.ModifyByParagraphBoundary`.
Instead of just moving backward or forward by paragraph granularity in the `UIAsyncTextInput` method
`-extendInDirection:byGranularity:`, we need to additionally move backwards or forwards by character
before doing so in order to maintain parity with `-_moveToEndOfParagraph:withHistory:` and
`_moveToStartOfParagraph:withHistory:`.

* Source/WebKit/UIProcess/ios/WKSEDefinitions.h: Renamed from Source/WebKit/UIProcess/API/ios/WKSEDefinitions.h.

Also move `WKSEDefinitions.h` to `UIProcess/ios` instead of `UIProcess/API/ios`, since the latter
just holds iOS-specific files related to WebKit API objects (in this case, only `WKWebView`).
Putting this in `UIProcess/ios` seems more appropriate, and also allows us to include it directly in
WebKitTestRunner.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestRunnerShared/spi/UIKitSPIForTesting.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
(-[TestWKWebView synchronouslyAdjustSelectionWithDelta:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ImageAnalysisTests.mm:
(TestWebKitAPI::simulateEditContextMenuAppearance):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKContentViewEditingActions.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEditActions.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/iOSMouseSupport.mm:
(TestWebKitAPI::simulateEditContextMenuAppearance):
* Tools/TestWebKitAPI/Tests/ios/AutocorrectionTestsIOS.mm:
(-[TestWKWebView autocorrectionRectsForString:]): Deleted.
(-[TestWKWebView synchronouslyRequestAutocorrectionContext]): Deleted.
* Tools/TestWebKitAPI/Tests/ios/EnterKeyHintTests.mm:
(-[TestWKWebView test:enterKeyHint:returnKeyType:]):
* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
* Tools/TestWebKitAPI/Tests/ios/SelectionModifyByParagraphBoundary.mm:
* Tools/TestWebKitAPI/Tests/ios/TextServicesTests.mm:
* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:

Add a bunch of new category helpers on `WKWebView`, which replace many of the direct calls into
`WKContentView` in our API tests when exercising various `UIWKInteractionViewProtocol` SPIs. These
helper methods will call directly into these legacy SPIs when async text input is disabled, but will
call into their new `UIAsyncTextInput` counterparts when async text input is enabled.

(-[WKWebView defineSelection]):
(-[WKWebView shareSelection]):
(-[WKWebView hasAsyncTextInput]):
(-[WKWebView selectionClipRect]):
(-[WKWebView moveSelectionToStartOfParagraph]):
(-[WKWebView extendSelectionToStartOfParagraph]):
(-[WKWebView moveSelectionToEndOfParagraph]):
(-[WKWebView extendSelectionToEndOfParagraph]):
(-[WKWebView asyncTextInput]):
(-[WKWebView extendedTextInputTraits]):
(-[WKWebView autocorrectionRectsForString:]):
(-[WKWebView autocorrectionContext]):
(-[WKWebView effectiveTextInputTraits]):
(-[WKWebView replaceText:withText:shouldUnderline:completion:]):
(wrap):
(unwrap):
(-[WKWebView synchronouslyAdjustSelectionWithDelta:]):
(-[WKWebView selectTextForContextMenuWithLocationInView:completion:]):
(-[WKWebView handleKeyEvent:completion:]):
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::asyncTextInput const):

Canonical link: https://commits.webkit.org/272233@main
